### ISSUE TITLE
Add FilePath.descendant(segmentsToChild) to replace fp.child(a).child(b).child(c)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3169]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#3169|
|Type|enhancement|
|Created|2008-04-06 02:45:46Z|
|Branch|https://github.com/twisted/twisted/tree/child-args-3169-2|

`fp.descendant([a, b])` should be equivalent to `fp.child(a).child(b)`.

Attachments:

* [child-changes.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/fe8/fe85b11fb29ae3cf6ca9eaaa025972868e628b58/2c9ba54361776dc6253e32cbebc726e0a1b53135.patch) (3632 bytes) - added by exarkun on 2008-11-01 01:35:39Z - examples of updating FilePath.child-using APIs (not part of the proposed changeset)

<details><summary>Searchable metadata</summary>

```
trac-id__3169 3169
type__enhancement enhancement
reporter__radix radix
priority__normal normal
milestone__ 
branch__branches_child_args_3169_2 branches/child-args-3169-2
branch_author__exarkun__radix exarkun, radix
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1207449946000000 1207449946000000
changetime__1279043893000000 1279043893000000
version__None None
owner__ 
cc__exarkun cc__zseil
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_1"></a>|@radix commented|
|-|-|

(In [23211]) Branching to 'child-args-3169'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Why?

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_3"></a>|@radix commented|
|-|-|

Convenience. I really do type .child('x').child('y').child('z') a *lot*.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

A different suggestion for this use-case: a different method, say, named "descend" that takes a list.  It would be nice to be able to test for existence of this convenience feature, and it would possibly be nice to pass around lists without using x(*y) syntax...

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Seems like 6 of 1...

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_6"></a>|@radix commented|
|-|-|

I could probably offer some arguments for .child(*x) instead of .descend(l). Let's see.

I think if we extend it to .child(*a), it would naturally add convenience without burdening implementors with writing a new method, and I assume that's why you suggested the desire to check for its existence. Implementing *a is quite straightforward (I've already done it for FilePath and partially for ZipPath in the branch, just need to go polish it off and update docstrings).

From a user's perspective, child(a) very naturally extends to child(a, b). It doesn't require going and learning about a new method and remembering its name. I don't think it's a problem to use *l syntax when you have a list.

So while the feature is intended for convenience, I think doing it this way feels way better than adding a "convenience method", as such.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Replying to [radix](#note_6):
> I could probably offer some arguments for .child(*x) instead of .descend(l). Let's see.
>
> I think if we extend it to .child(*a), it would naturally add convenience without burdening implementors with writing a new method, and I assume that's why you suggested the desire to check for its existence. Implementing *a is quite straightforward (I've already done it for FilePath and partially for ZipPath in the branch, just need to go polish it off and update docstrings).

"burdening implementors with writing a new method"?  I don't know dvorak - is it really so hard to type 'def' on your keyboard?  I only have to use two fingers; maybe you should switch to qwerty.

Seriously though, I don't see "*a" vs. a new method as particularly burdensome for the implementor.

> From a user's perspective, child(a) very naturally extends to child(a, b). It doesn't require going and learning about a new method and remembering its name. I don't think it's a problem to use *l syntax when you have a list.

Similarly my argument about * syntax for lists was probably kind of bogus.  It's just evolved into a code smell for me: when I start typing "*[" or "**{" a lot, I start thinking that I mis-designed an API; possibly not appropriate here.   But the unpleasantness of debugging why your code works except for that one place where it uses the convenience syntax, or looking at `fp.child.im_func.func_code.co_argcount` or whatever is real.

> So while the feature is intended for convenience, I think doing it this way feels way better than adding a "convenience method", as such.

It "feels" worse to me, but this should all be taken as advice, not a review or a roadblock.  Add it if you think it's good.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [25241]) FilePath.child(*path)

refs [#3169](https://github.com/twisted/twisted/issues/3169)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun removed owner|
|-|-|

I started on this before I noticed the discussion.  Since I didn't see any changes in the branch, I went ahead and checked in what I wrote.

I'm somewhat ambivalent about `*args` vs a new method.  I think the implementation ends up being a bit easier to read with a new method (since there's some much junk in `FilePath.child`).  However, it does seem to me like `p.child(a, b, c)` is a more obvious thing to try/want than `p.anothermethod([a, b, c])`.  On the other hand, `*args` APIs are weird.  Anyway whatever.  I can't actually think of a new good method name for what this ticket wants, and I already wrote the `*args` version.

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_10"></a>|@zseil set owner to @exarkun|
|-|-|

Here is a quick review:
 1. In [FilePath.child()](http://twistedmatrix.com/trac/browser/branches/child-args-3169/twisted/python/filepath.py#L310), `segment.count(":")` should be probably changed to `":" in segment`.
 2. I noticed that `path.child("sub1", "..", "sub2")` is not allowed by the current code. If this is intentional, it should probably have a test.
 3. `ZipPath.child()` needs a docstring.
 3. The signature change is incompatible with `twisted.python.urlpath.URLPath.child()`'s signature. I don't know if this is a problem or not.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [29503]) Branching to 'child-args-3169-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

I changed the implementation to `FilePath.descendant` (and ZipPath too).  This shortens the diff quite a bit, at least.  I am mainly convinced by the `fp.child(*foo)` argument.

I continued not to worry about `URLPath`, because `URLPath` only _wants_ to be the same kind of thing as `FilePath`, it doesn't actually succeed, though hopefully someday it might.

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/child-args-3169-2)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

update ticket description to match what's implemented

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set owner to @exarkun|
|-|-|

Hooray, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [29803]) Merge child-args-3169-2

Author: exarkun
Reviewer: glyph, zseil, therve
Fixes: [#3169](https://github.com/twisted/twisted/issues/3169)

Add `FilePath.descendant` to replace repeated `child` calls.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
